tools/c7n_azure/c7n_azure/resources/network_security_group.py

```diff
@@ -0,0 +1,207 @@
+from c7n_azure.provider import resources
+from c7n.actions import BaseAction
+from c7n.filters import Filter, FilterValidationError
+from c7n.utils import type_schema
+
+
+@resources.register('networksecuritygroup')
+class NetworkSecurityGroup(QueryResourceManager):
+    class resource_type(object):
+        service = 'azure.mgmt.network'
+        client = 'NetworkManagementClient'
+        enum_spec = ('network_security_groups', 'list_all')
+        id = 'id'
+        name = 'name'
+        default_report_fields = (
+            'name',
+            'location',
+            'resourceGroup'
+        )
+
+
+class SecurityRuleFilter(Filter):
+    perm_attrs = set((
+        'IpProtocol', 'FromPort', 'ToPort', 'UserIdGroupPairs',
+        'IpRanges', 'PrefixListIds'
+    ))
+
+    filter_attrs = set(('Cidr', 'Ports', 'OnlyPorts', 'SelfReference'))
+    attrs = perm_attrs.union(filter_attrs)
+    attrs.add('match-operator')
+
+    def process(self, network_security_groups, event=None):
+        self.ip_protocol = 'IpProtocol' in self.data and self.data['IpProtocol'] or None
+        self.from_port = 'FromPort' in self.data and self.data['FromPort'] or None
+        self.to_port = 'ToPort' in self.data and self.data['ToPort'] or None
+        self.ports = 'Ports' in self.data and self.data['Ports'] or None
+        self.only_ports = (
+                'OnlyPorts' in self.data and self.data['OnlyPorts'] or None)
+        if self.from_port and self.to_port and self.from_port > self.to_port:
+            raise ValueError('FromPort should be lower than ToPort')
+        if ((self.from_port or self.to_port) and (self.ports or self.only_ports)) or (self.ports and self.only_ports):
+            raise ValueError('Invalid port parameters. Choose port range (FromPort and/or ToPort) or '
+                             'specify specific ports (Ports or OnlyPorts)')
+        match_op = self.data.get('match-operator', 'and') == 'and' and all or any
+        for nsg in network_security_groups:
+            nsg['properties']['securityRules'] = \
+                [rule for rule in nsg['properties']['securityRules'] if self.is_match(rule, match_op)]
+        network_security_groups = \
+            [nsg for nsg in network_security_groups if len(nsg['properties']['securityRules']) > 0]
+        return network_security_groups
+
+    def is_range_match(self, dest_port_range):
+        # destination port range is coming from Azure, existing rules, not policy input
+        if len(dest_port_range) > 2:
+            raise ValueError('Invalid range')
+        # Both FromPort and ToPort are specified, should fall within range
+        if self.from_port and self.to_port:
+            for port in dest_port_range:
+                if port > self.to_port or port < self.from_port:
+                    return False
+        # Just FromPort is specified, should be above FromPort
+        elif self.from_port:
+            for port in dest_port_range:
```

Comment:
if this is a range, i.e. two values, you should only be checking the first one, dest_port_range[0] rather than both. ideally the port is < dest_port_range[1] in this block
```python
if self.from_port and self.from_port < dest_port_range[0]:
  return False
```

same idea in line 84

Reply:
currently the schema would allow it from what I can see. I don't think there's a particularly obvious reason why we shouldn't